Network traffic monitoring

jukkar opened the pull request:
This is useful in debugging connectivity issues and/or new protocol development. The idea is to capture sent or received network traffic and send it to external system for postprocessing / viewing with Wireshark or similar tool.

jukkar committed:
net: l2: ipip: Add capability bit for IPIP tunnel

The IPIP tunnel capability bit helps to detect which network
interface supports IPIP tunneling.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: capture: Add support for network packet capturing

Add infrastructure to allow user to configure the system so that
all the network packets, that are sent to or received from a specific
network interface, are sent to remote system for analysis.
The captured network packets are placed as a payload in UDP packet,
which is then sent inside a tunnel to a remote host. The host can
then receive the packets and for example show them in wireshark.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Make net_pkt_clone() to use the original slab

Instead of forcing net_pkt_clone() use the tx_pkts slab, use
the same slab that was used when allocating the original net_pkt.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Add network packet capture support

Add commands in shell to setup and enable/disable network
packet capture.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Show detailed interface information

Show detailed information about network interface that is down,
only when user asks such information about one specific network
interface. This means that for "net iface" command only
"Interface is down." is printed but the command "net iface 1"
will print detailed information even if interface is down.
This helps to view the information and user does not need
to bring interface up to see the details.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Add helper to store capture bit

Add get/set helpers to store information whether the net_pkt
is captured already or not. This bit is used to detect capture
loop and avoid recursion.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: l2: ipip: Support more than one IPIP tunnel

Alloc code to create more than one IPIP tunnel interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: capture: Catch sent and received packets

Create net_l2_send() function which will be called by each L2
sending function so that we can catch all the network packets
that are being sent. Some L2 layers send things a bit differently,
so in those cases call the net_capture_send() directly by the L2
layer.
Add network packet capture call in receive side after the pkt has
been received by the RX queue handler. This avoids calling the
net_capture_send() from ISR context.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: capture: Network packet capture app

A simple application that allows you to use net-shell to
configure network packet capturing. The captured packets are
sent to external systems for processing. This can be used
for debugging network protocols.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: doc: Add network monitoring documentation

Add information to documentation how to monitor and capture
received and sent network traffic.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print external net_pkt slab information

The "net mem" command did not print external net_pkt slabs
properly (the number of free net_pkt's was not printed).

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

rlubos reviewed (commented):
Spotted a few nits, but it looks pretty good in general.

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -8,8 +8,8 @@
 #include <net/net_core.h>
-#include <net/net_l2.h>
 #include <net/net_if.h>
+#include <net/net_l2.h>
```
Yep, by tweaking capture.h a bit I managed to get rid of this commit.

jukkar committed:
modules: net-tools: Use updated tunneling conf file

Use the fixed tunneling zeth-tunnel.conf file from net-tools
project.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: echo-server: Add capture overlay file

Add an overlay-capture.conf file that can be used to allow network
packet capturing in this sample app.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
* update to latest upstream
* documentation updates
* added capture calls to various networking L2 components

chrta reviewed (changes requested):
Just one minor typo. Thanks for implementing this, it is a really cool feature.

mniestroj commented on subsys/net/lib/capture/Kconfig:
```diff
@@ -0,0 +1,95 @@
+	  if one needs to send captured data to multiple different devices,
+	  then you need to increase the value.
+
+config NET_CAPTURE_TUNNEL_REMOTE_ADDR
```
This option seems dead, it is never used in the code.

mniestroj commented on subsys/net/ip/net_shell.c:
```diff
@@ -1754,6 +1756,241 @@ static int cmd_net_arp_flush(const struct shell *shell, size_t argc,
+	int ret, arg = 1;
+	const char *remote, *local, *peer;
+
+	if (capture_dev != NULL) {
+		PR_INFO("Capture already setup. "
+			"Do \"net capture cleanup\" first.\n");
+		return 0;
+	}
```
Can't we just "cleanup" automatically instead of requiring user to call `net capture cleanup` manually?

jukkar replied:
I was thinking this but for now decided to go this route. The API calls have now symmetry, so there is setup() which has cleanup() counterpart, and then enable() and disable().

jukkar replied:
Ah, I actually misunderstood you. Sure, we can have the shell do cleanup automatically.

mniestroj commented on doc/guides/networking/network_monitoring.rst:
```diff
@@ -0,0 +1,222 @@
+It is useful to be able to monitor the network traffic especially when
+debugging a connectivity issues or when developing new protocol support in
+Zephyr. This page describes how to set up a way to capture network traffic so
+that user is able to use Wireshark or similar tool in remote host so see the
```
`s/so see the/to see the/` ?

jukkar commented on subsys/net/lib/capture/Kconfig:
```diff
@@ -0,0 +1,95 @@
+	  192.0.2.2
+	  2001:db8::2
+
+config NET_CAPTURE_TUNNEL_MY_ADDR_PORT
```
Yes, this was done on purpose. The use is / should be explained on the documentation commit. I created this option so that application can use it and do not need to create the Kconfig entry for this purpose.

mniestroj replied:
But then sample application configures those options and there is no visible change to the user. It might be confusing when user will configure those and still don't get automatically configured tunnel. If they are defined here, then lets at least use those options in sample application. Otherwise I don't see the reason to keep them if they are just "ghost" options and never used.

jukkar replied:
The documentation has now this
```
It is not necessary to use ``net-shell`` for configuring the monitoring.
The network capture API functions can be called by the application if needed.
For that purpose, the capture Kconfig file contains these configure options

.. csv-table::
   :widths: auto

   :option:`CONFIG_NET_CAPTURE_TUNNEL_REMOTE_ADDR`
   :option:`CONFIG_NET_CAPTURE_TUNNEL_MY_ADDR_PORT`
   :option:`CONFIG_NET_CAPTURE_TUNNEL_PEER_ADDR_PORT`

which can be set by the application ``prj.conf`` file and then given as
parameters when calling :c:func:`net_capture_setup` function.
These Kconfig options are currently not used by the capture sample
application or by the net_capture API but are provided as a convenience for
the end user.
```
but perhaps it is confusing to the reader. The motivation for doing it like this was that we could have consistent naming for the config options even if they are not directly used by the service itself.

jukkar replied:
I will remove the extra options as they can be confusing to the end user.

jukkar commented:
Updated according to comments.

chrta reviewed (approved):
LGTM, but i have not tested this.
